### PR TITLE
fix: warn when implicit label has no visible text

### LINKED ISSUE #1860: input-requires-label: warn when label wraps input but contains no visible text
**Is your feature request related to a problem? Please describe.**
After #1852 landed implicit label support, `input-requires-label` now correctly allows `<input>` nested inside `<label>`. However, it doesn't check whether the label actually contains visible text. This passes without warning:

`<label><input type="password" /></label>`

  This is not accessible as a screen reader finds the input but has no name to announce. (Raised by @coliff in
  https://github.com/htmlhint/HTMLHint/pull/1852#issuecomment-4264523894.)

**Describe the solution you'd like**
The rule should warn when an implicit `<label>` wraps an `<input>` but contains no non-whitespace text content.
Should warn:
`<label><input type="password" /></label>`
`<label>   <input type="text" />   </label>`

Should pass:
`<label>Password <input type="password" /></label>`
`<label><span>Email</span> <input type="email" /></label>`

**Proposed approach**
The parser already emits a `text` event for text nodes between tags, and `h1-require` uses this exact pattern to check for empty `<h1>` tags. The same approach applies here:
1. When `<label>` (without `for`) opens, reset a `labelHasText` flag to `false`
2. On the `text` event, if inside an implicit label and the text isn't whitespace-only, set `labelHasText = true`
3. When `</label>` closes, if the label contained inputs but `labelHasText` is still `false`, report a warning

This reuses the `labelDepth` tracking from #1852 and adds one event listener + one boolean flag.

**Questions before I start**
1. Should this live in `input-requires-label`, or would you prefer a separate rule like `label-requires-text`?
2. Should `<label>icon <input></label>` (no text node, but `alt` text on a child element) warn or pass? I'm leaning toward warn, since `alt` isn't visible label text, but wanted to check.

Happy to submit a PR for this.

### PR DESCRIPTION
**Pull request by ShivaniNR**
## Short description of what this resolves
Closes #1860

An implicit label wrapping an input (`<label><input /></label>`) without any visible text is inaccessible; screen readers find the input but have no name to announce. After #1852 added implicit label support, this case was silently accepted.

## Proposed changes
- Added a `labelHasText` flag that resets when an implicit `<label>` opens
- Added a `text` event listener that sets the flag when non-whitespace text is found inside the label
- On `</label>` close, if no text was found, nested inputs are marked as unlabeled, so the existing warning fires
- Updated tests: empty/whitespace-only labels now warn, labels with visible text pass

### CODE REVIEW

File: src/core/rules/input-requires-label.ts [hunk 1]
```diff
@@ -17,6 +17,7 @@ export default {
       nested: boolean
     }> = []
     let labelDepth = 0
+    let labelHasText = false
```
[the remark this answers is not in the document]
Comment (ShivaniNR) [reply src/core/rules/input-requires-label.ts]: Done. Replaced `let labelHasText = false` with const `labelStateStack: Array<{ hasText: boolean; inputStartIndex: number }> = []` to track state per label depth

File: src/core/rules/input-requires-label.ts [hunk 2]
```diff
@@ -43,16 +44,30 @@ export default {
           // implicit label (no `for`): nesting labels the input
           // a self-closing <label/> opens no scope and emits no tagend
           labelDepth++
+          labelHasText = false
```
[the remark this answers is not in the document]
Comment (ShivaniNR) [reply src/core/rules/input-requires-label.ts]: Done. Now pushing `{ hasText: false, inputStartIndex: inputTags.length }` on label open instead of resetting a global boolean.

File: src/core/rules/input-requires-label.ts [hunk 3]
```diff
@@ -43,16 +44,30 @@ export default {
...
     parser.addListener('tagend', (event) => {
       if (event.tagName.toLowerCase() === 'label' && labelDepth > 0) {
+        if (!labelHasText) {
+          inputTags.forEach((input) => {
+            if (input.nested) {
+              input.nested = false
+            }
+          })
+        }
         labelDepth--
```
[the remark this answers is not in the document]
Comment (ShivaniNR) [reply src/core/rules/input-requires-label.ts]: Done. The `tagend` listener now pops the stack, only resets `nested` for inputs within the current label's scope (from `inputStartIndex`), and propagates `hasText` to the parent label. Also added two tests covering the multi-label and nested-label edge cases.

File: src/core/rules/input-requires-label.ts [hunk 4]
```diff
@@ -43,16 +44,30 @@ export default {
...
       }
     })
 
+    parser.addListener('text', (event) => {
+      if (labelDepth > 0 && event.raw && !/^\s*$/.test(event.raw)) {
+        labelHasText = true
```
[the remark this answers is not in the document]
Comment (ShivaniNR) [reply src/core/rules/input-requires-label.ts]: Done.

File: dist/core/rules/input-requires-label.js [hunk 5]
```diff
@@ -7,6 +7,7 @@ exports.default = {
         const labelTags = [];
         const inputTags = [];
         let labelDepth = 0;
+        let labelHasText = false;
```
[the remark this answers is not in the document]
Comment (ShivaniNR) [reply dist/core/rules/input-requires-label.js]: Done.

File: dist/core/rules/input-requires-label.js [hunk 6]
```diff
@@ -27,14 +28,27 @@ exports.default = {
                 }
                 else if (!event.close) {
                     labelDepth++;
+                    labelHasText = false;
```
[the remark this answers is not in the document]
Comment (ShivaniNR) [reply dist/core/rules/input-requires-label.js]: Done.

File: dist/core/rules/input-requires-label.js [hunk 7]
```diff
@@ -27,14 +28,27 @@ exports.default = {
...
         parser.addListener('tagend', (event) => {
             if (event.tagName.toLowerCase() === 'label' && labelDepth > 0) {
+                if (!labelHasText) {
+                    inputTags.forEach((input) => {
+                        if (input.nested) {
+                            input.nested = false;
+                        }
+                    });
+                }
                 labelDepth--;
```
[the remark this answers is not in the document]
Comment (ShivaniNR) [reply dist/core/rules/input-requires-label.js]: Done.

File: dist/core/rules/input-requires-label.js [hunk 8]
```diff
@@ -27,14 +28,27 @@ exports.default = {
...
             }
         });
+        parser.addListener('text', (event) => {
+            if (labelDepth > 0 && event.raw && !/^\s*$/.test(event.raw)) {
+                labelHasText = true;
```
[the remark this answers is not in the document]
Comment (ShivaniNR) [reply dist/core/rules/input-requires-label.js]: Done.

**Comment by coliff** [comment]
PR looking really good. thanks for your work on this! Just one comment left from Copilot to review I think.

File: test/rules/input-requires-label.spec.js [hunk 9]
```diff
@@ -61,6 +69,20 @@ describe(`Rules: ${ruleId}`, () => {
...
+      const messages = HTMLHint.verify(code, ruleOptions)
+      expect(messages.length).toBe(1)
+    })
+
+    it('Nested labels: outer with text should not false-warn outer input', () => {
+      const code =
+        '<label>Outer <label><input type="checkbox" /></label><input type="text" /></label>'
+      const messages = HTMLHint.verify(code, ruleOptions)
+      expect(messages.length).toBe(1)
+    })
```
[the remark this answers is not in the document]
Comment (ShivaniNR) [reply test/rules/input-requires-label.spec.js]: Done. Moved both tests to the Error cases block since they expect warnings.

**Comment by ShivaniNR** [comment]
> PR looking really good. thanks for your work on this! Just one comment left from Copilot to review I think.

Thanks! Moved the two tests to the Error cases block as Copilot suggested.

**Review by coliff — APPROVED** [review]
